subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java

```diff
@@ -226,53 +159,36 @@ public void validate(String propertyName, boolean optional, ValidationAction val
         }
     }
 
-    private static class ImplementationPropertyValue implements ValidatingValue {
-
-        private final Class<?> beanClass;
-
-        public ImplementationPropertyValue(Class<?> beanClass) {
-            this.beanClass = beanClass;
-        }
-
-        @Override
-        public Object call() {
-            return beanClass;
-        }
-
-        @Override
-        public void validate(String propertyName, boolean optional, ValidationAction valueValidator, TaskValidationContext context) {
-        }
-    }
-
-    private static class AbsentPropertyValue implements ValidatingValue {
-        @Nullable
-        @Override
-        public Object call() {
-            return null;
+    private class NestedPropertyContext extends AbstractNestedBeanContext<PropertyNode> {
+        private final PropertyNode currentNode;
+        private final Queue<NestedPropertyContext> queue;
+        private final Iterable<PropertyNode> parentNodes;
+
+        public NestedPropertyContext(PropertyNode currentNode, Queue<NestedPropertyContext> queue, Iterable<PropertyNode> parentNodes) {
+            super(propertyMetadataStore);
+            this.currentNode = currentNode;
+            this.queue = queue;
+            this.parentNodes = parentNodes;
+            for (PropertyNode parentNode : parentNodes) {
+                Preconditions.checkState(
+                    currentNode.getBean() != parentNode.getBean(),
+                    "Cycles between nested beans are not allowed. Cycle detected between: '%s' and '%s'.",
+                    parentNode, currentNode);
+            }
         }
 
         @Override
-        public void validate(String propertyName, boolean optional, ValidationAction valueValidator, TaskValidationContext context) {
-            context.recordValidationMessage(ERROR, String.format("No value has been specified for property '%s'.", propertyName));
-        }
-    }
-
-    private static class InvalidPropertyValue implements ValidatingValue {
-        private final Exception exception;
-
-        public InvalidPropertyValue(Exception exception) {
-            this.exception = exception;
+        public PropertyNode createNode(String propertyName, Object nested) {
+            return new PropertyNode(propertyName, nested);
         }
 
-        @Nullable
         @Override
-        public Object call() {
-            return null;
+        public void addNested(PropertyNode node) {
+            queue.add(new NestedPropertyContext(node, queue, Iterables.concat(ImmutableList.of(currentNode), parentNodes)));
```

Comment:
This could be somewhat more frugal with resources: https://gist.github.com/lptr/d4d45c8b94b29a08d8989644c540b7ec

Reply:
Changed it to your code.